pkg/api/server/v1alpha2/lister/aggregator.go

```diff
@@ -0,0 +1,208 @@
+		equalityClauses: clauses,
+	}
+
+	summary := strings.Split(strings.TrimSpace(aggregateObjectRequest.GetSummary()), ",")
+	if len(summary) == 0 {
+		// make total default
+		summary = append(summary, "total")
+	}
+
+	aggregators, err := getAggregateFunc(summary)
+	if err != nil {
+		return nil, err
+	}
+
+	group := strings.TrimSpace(aggregateObjectRequest.GetGroupBy())
+	if group != "" {
+		_, ok := validGroups[group]
+		if !ok {
+			return nil, status.Errorf(codes.InvalidArgument, "group_by does not recognize %s", group)
+		}
+		aggregators = append(aggregators, groupBy(group, groupQuery))
+	}
+
+	return &Aggregator{
+		aggregators: aggregators,
+		queryBuilders: []queryBuilder{
+			filter,
+		},
+	}, nil
+}
+
+func getDuration(fn, query, value string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Select(db.Statement.Selects, fn+"("+query+")::INTERVAL as "+value)
+	}
+}
+
+func groupBy(group, query string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		groupSelect := "EXTRACT(EPOCH FROM DATE_TRUNC('" + group + "', " + query + ")) AS group_value"
+		return db.Select(db.Statement.Selects, groupSelect).Group("group_value")
+	}
+}
+
+func getStatus(query, reason string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		statusSelect := ""
+		if reason == "Unknown" {
+			statusSelect = "COUNT(CASE WHEN " + query + " NOT IN ('Failed', 'Succeeded', 'Cancelled') THEN 1 END) AS unknown"
+		} else {
+			statusSelect = "COUNT(CASE WHEN " + query + " = '" + reason + "' THEN 1 END) AS " + strings.ToLower(reason)
+		}
+		return db.Select(db.Statement.Selects, statusSelect)
+	}
+}
+
+func getCount(query, countName string) aggregateFunc {
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Select(db.Statement.Selects, "COUNT("+query+") AS "+countName)
+	}
+}
+
+type aggregateFunc func(db *gorm.DB) *gorm.DB
+
+var summaryFuncs = map[string]aggregateFunc{
+	"total":          getCount("*", "total"),
+	"avg_duration":   getDuration("AVG", durationQuery, "avg_duration"),
+	"max_duration":   getDuration("MAX", durationQuery, "max_duration"),
+	"total_duration": getDuration("SUM", durationQuery, "total_duration"),
+	"min_duration":   getDuration("MIN", durationQuery, "min_duration"),
+	"succeeded":      getStatus(statusQuery, "Succeeded"),
+	"failed":         getStatus(statusQuery, "Failed"),
+	"cancelled":      getStatus(statusQuery, "Cancelled"),
+	"completed":      getStatus(statusQuery, "Completed"),
+	"unknown":        getStatus(statusQuery, "Unknown"),
+}
+
+func getAggregateFunc(queries []string) ([]aggregateFunc, error) {
+	fns := make([]aggregateFunc, 0, len(queries))
+	for _, q := range queries {
+		fn, ok := summaryFuncs[q]
+		if !ok {
+			return nil, status.Errorf(codes.InvalidArgument, "invalid aggregate query: %s", q)
+		}
+		fns = append(fns, fn)
+	}
+	return fns, nil
+}
+
+func (a *Aggregator) applyAggregateFunc(ctx context.Context, db *gorm.DB) *gorm.DB {
```

Comment:
In a lot of these methods we are already sending the `gorm.DB` struct as pointer and return the same pointer.

Reply:
yes, I do see an improvement scope here by including this as a part of the struct. Will do that.